hw/arm/armtrickbox.c

```diff
@@ -29,68 +29,229 @@
 #include "hw/sysbus.h"
+#include "hw/arm/armtrickbox.h"
 #include "qemu/module.h"
 #include "qapi/error.h"
 #include "hw/qdev-properties.h"
 #include "qemu/qemu-print.h"
 
 #define BUF_SIZE 100
 
+// 0: none. 1: warn. 2: info
+#define TRICKBOX_LOG_LEVEL 2
+
+// It is quite useful when debugging to stop tracing on first failure. Just set
+// this to an appropriate string.
+const char *const exit_string = NULL; // "FAIL";
+
+#define LOG_ERROR 0
+#define LOG_WARN  1
+#define LOG_INFO  2
+static const char *const levels[] = { "ERROR:", "WARN:", "INFO:" };
+
+#define TRICKBOX_LOG(lvl, fmt, ...)                                            \
+    if (TRICKBOX_LOG_LEVEL >= lvl)                                             \
+    printf("TRICKBOX: %s" fmt, levels[lvl], ##__VA_ARGS__)
+
+/*
+ * List of registers to handle just by getting/setting a field in
+ * arm_trickbox_state. Also optionally call a helper after any update.
+ * ITEM(name, reset_value, read_mask, write_mask, write_helper, [alias])
+ */
+
+#define REGISTER_LIST(ITEM)                                                    \
+    ITEM(Configuration, 0x00004002, 0xFFFFFFFE, 0xFFFFFFFF, NULL)              \
+    ITEM(TargetCPU, 0xF, 0xF, 0xF, NULL)                                       \
+    ITEM(abortrgnlo1, 0x40000000, 0xFFFFFFFF, 0xFFFFFFFF,                      \
+         axi_abort_range_changed, AbortRgn64Lo1)                               \
+    ITEM(abortrgnhi1, 0x50000000, 0xFFFFFFFF, 0xFFFFFFFF,                      \
+         axi_abort_range_changed, AbortRgn64Hi1)                               \
+    ITEM(abortrgnlo2, 0x3F000000, 0xFFFFFFFF, 0xFFFFFFFF,                      \
+         axi_abort_range_changed, AbortRgn64Lo2)                               \
+    ITEM(abortrgnhi2, 0x3F080000, 0xFFFFFFFF, 0xFFFFFFFF,                      \
+         axi_abort_range_changed, AbortRgn64Hi2)                               \
+    ITEM(AbortRgn64Lo1_Hi, 0x0, 0xFFFFFFFF, 0xFFFFFFFF,                        \
+         axi_abort_range_changed)                                              \
+    ITEM(AbortRgn64Hi1_Hi, 0x0, 0xFFFFFFFF, 0xFFFFFFFF,                        \
+         axi_abort_range_changed)                                              \
+    ITEM(AbortRgn64Lo2_Hi, 0x0, 0xFFFFFFFF, 0xFFFFFFFF,                        \
+         axi_abort_range_changed)                                              \
+    ITEM(AbortRgn64Hi2_Hi, 0x0, 0xFFFFFFFF, 0xFFFFFFFF,                        \
+         axi_abort_range_changed)                                              \
+    ITEM(AXIAbortCtl, 0x0, 0xFFFFFFFF, 0xFFFFFFFF, axi_abort_changed)          \
+    ITEM(DBGControl, 0x00000181, 0xFFFFFFFF, 0xC1, NULL)
+
+// A stripped down MAP that only applies to 2, 1 or 0 elements.
+#define NONE(...)
+#define ONE(F, X)               F(X)
+#define TWO(F, X, Y)            F(X) F(Y)
+#define SELECT(F, A, B, C, ...) C
+#define MAP(...)                SELECT(__VA_ARGS__, TWO, ONE, NONE)(__VA_ARGS__)
```

Comment:
You don't need kludgy MAP stuff if you just make your inner alias lists be themselves macro lists in the same style, i.e.

```
#define LIST(ITEM, ALIAS) \
    ITEM(foo, ALIAS(bar))
/* ... */
#define CASE_HANDLE(name) ...
#define HANDLE_WRITE(foo, ...) \
    CASE_HANDLE(foo) \
    __VA_ARGS__ \
        ... \
        break;
LIST(HANDLE_WRITE, CASE_HANDLE)
```
kind of thing (and that then avoids any arbitrary limits)